In [207]:
# Підгружаю необхідні бібліотеки
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

## Підготовка даних

In [208]:
# Зчитую файл
url = "https://docs.google.com/spreadsheets/d/1mAP0_Noi8f9-Ha4et5TlksabocHsPtUM7EalYH5xX_o/export?format=csv&gid=0"

# ПРибираю назви стовбців, шоб побачити де закінчується перша таблиця
data = pd.read_csv(url, header=None)
data

0         1                 2                            3   4  \
0     user_id  is_trial  first_event_date  subscription_renewal_amount NaN   
1       21889         1            1/1/22                            0 NaN   
2       21890         1            7/1/19                            0 NaN   
3       21891         1            7/1/19                            0 NaN   
4       21892         1            7/1/19                            0 NaN   
...       ...       ...               ...                          ...  ..   
9422    23472         0            7/6/19                            3 NaN   
9423    22991         0            7/5/19                            7 NaN   
9424    23248         0            7/5/19                           11 NaN   
9425    24355         0            7/7/19                           23 NaN   
9426    24769         1            7/7/19                            1 NaN   

                                                5             6  \
0                                       Subs type  Trial period   
1                                           trial        7 days   
2                                        wo trial      No trial   
3     Subscription is renewed automatically\t\t\t           NaN   
4                                             NaN           NaN   
...                                           ...           ...   
9422                                          NaN           NaN   
9423                                          NaN           NaN   
9424                                          NaN           NaN   
9425                                          NaN           NaN   
9426                                          NaN           NaN   

                         7              8               9  
0     Subscription period   First payment  Rebill payment  
1                  1 month          $6.99          $29.99  
2                3 months          $40.00          $40.00  
3                      NaN            NaN             NaN  
4                      NaN            NaN             NaN  
...                    ...            ...             ...  
9422                   NaN            NaN             NaN  
9423                   NaN            NaN             NaN  
9424                   NaN            NaN             NaN  
9425                   NaN            NaN             NaN  
9426                   NaN            NaN             NaN  

[9427 rows x 10 columns]

In [209]:
# Видаляю 4 стовбикю, він пустий
data = data.drop(columns=[4])
data

0         1                 2                            3  \
0     user_id  is_trial  first_event_date  subscription_renewal_amount   
1       21889         1            1/1/22                            0   
2       21890         1            7/1/19                            0   
3       21891         1            7/1/19                            0   
4       21892         1            7/1/19                            0   
...       ...       ...               ...                          ...   
9422    23472         0            7/6/19                            3   
9423    22991         0            7/5/19                            7   
9424    23248         0            7/5/19                           11   
9425    24355         0            7/7/19                           23   
9426    24769         1            7/7/19                            1   

                                                5             6  \
0                                       Subs type  Trial period   
1                                           trial        7 days   
2                                        wo trial      No trial   
3     Subscription is renewed automatically\t\t\t           NaN   
4                                             NaN           NaN   
...                                           ...           ...   
9422                                          NaN           NaN   
9423                                          NaN           NaN   
9424                                          NaN           NaN   
9425                                          NaN           NaN   
9426                                          NaN           NaN   

                         7              8               9  
0     Subscription period   First payment  Rebill payment  
1                  1 month          $6.99          $29.99  
2                3 months          $40.00          $40.00  
3                      NaN            NaN             NaN  
4                      NaN            NaN             NaN  
...                    ...            ...             ...  
9422                   NaN            NaN             NaN  
9423                   NaN            NaN             NaN  
9424                   NaN            NaN             NaN  
9425                   NaN            NaN             NaN  
9426                   NaN            NaN             NaN  

[9427 rows x 9 columns]

In [210]:
# Копіюю датафрейм, змінюю назви стовбчиків як має бути, за допомогою слайсу видаляю рядок з назвами колонок
df = data.copy()
df.columns = df.iloc[0]
df = df[1:]
df

0    user_id is_trial first_event_date subscription_renewal_amount  \
1      21889        1           1/1/22                           0   
2      21890        1           7/1/19                           0   
3      21891        1           7/1/19                           0   
4      21892        1           7/1/19                           0   
5      21894        1           7/1/19                           0   
...      ...      ...              ...                         ...   
9422   23472        0           7/6/19                           3   
9423   22991        0           7/5/19                           7   
9424   23248        0           7/5/19                          11   
9425   24355        0           7/7/19                          23   
9426   24769        1           7/7/19                           1   

0                                       Subs type Trial period  \
1                                           trial       7 days   
2                                        wo trial     No trial   
3     Subscription is renewed automatically\t\t\t          NaN   
4                                             NaN          NaN   
5                                             NaN          NaN   
...                                           ...          ...   
9422                                          NaN          NaN   
9423                                          NaN          NaN   
9424                                          NaN          NaN   
9425                                          NaN          NaN   
9426                                          NaN          NaN   

0    Subscription period  First payment Rebill payment  
1                 1 month         $6.99         $29.99  
2               3 months         $40.00         $40.00  
3                     NaN           NaN            NaN  
4                     NaN           NaN            NaN  
5                     NaN           NaN            NaN  
...                   ...           ...            ...  
9422                  NaN           NaN            NaN  
9423                  NaN           NaN            NaN  
9424                  NaN           NaN            NaN  
9425                  NaN           NaN            NaN  
9426                  NaN           NaN            NaN  

[9426 rows x 9 columns]

In [211]:
# ВИдалимо порожні рядки і коментар
df.loc[df['Subs type'] == 'Subscription is renewed automatically\t\t\t', 'Subs type'] = np.nan
df = df.dropna(how="all")
df

0    user_id is_trial first_event_date subscription_renewal_amount Subs type  \
1      21889        1           1/1/22                           0     trial   
2      21890        1           7/1/19                           0  wo trial   
3      21891        1           7/1/19                           0       NaN   
4      21892        1           7/1/19                           0       NaN   
5      21894        1           7/1/19                           0       NaN   
...      ...      ...              ...                         ...       ...   
9422   23472        0           7/6/19                           3       NaN   
9423   22991        0           7/5/19                           7       NaN   
9424   23248        0           7/5/19                          11       NaN   
9425   24355        0           7/7/19                          23       NaN   
9426   24769        1           7/7/19                           1       NaN   

0    Trial period Subscription period  First payment Rebill payment  
1          7 days              1 month         $6.99         $29.99  
2        No trial            3 months         $40.00         $40.00  
3             NaN                  NaN           NaN            NaN  
4             NaN                  NaN           NaN            NaN  
5             NaN                  NaN           NaN            NaN  
...           ...                  ...           ...            ...  
9422          NaN                  NaN           NaN            NaN  
9423          NaN                  NaN           NaN            NaN  
9424          NaN                  NaN           NaN            NaN  
9425          NaN                  NaN           NaN            NaN  
9426          NaN                  NaN           NaN            NaN  

[9426 rows x 9 columns]

In [212]:
# Перевірю тип данних в таблиці
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9426 entries, 1 to 9426
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   user_id                      9426 non-null   object
 1   is_trial                     9426 non-null   object
 2   first_event_date             9426 non-null   object
 3   subscription_renewal_amount  9426 non-null   object
 4   Subs type                    2 non-null      object
 5   Trial period                 2 non-null      object
 6   Subscription period          2 non-null      object
 7   First payment                2 non-null      object
 8   Rebill payment               2 non-null      object
dtypes: object(9)
memory usage: 662.9+ KB


In [213]:
# ПРиводжу стовбчики до правильних типів даних
df[['user_id', 'subscription_renewal_amount']] = df[['user_id', 'subscription_renewal_amount']].astype(int)
df['first_event_date'] = pd.to_datetime(df['first_event_date'])
df['Trial period_days'] = df['Trial period'].str.extract('(\d+)').astype(float)
df['Subscription period_month'] = df['Subscription period '].str.extract('(\d+)').astype(float)
df.drop(columns=['Trial period', 'Subscription period '], inplace=True)
df['First payment_doll'] = df['First payment'].str.replace('$', '', regex=False).astype(float)
df['Rebill payment__doll'] = df['Rebill payment'].str.replace('$', '', regex=False).astype(float)
df.drop(columns=['First payment', 'Rebill payment'], inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9426 entries, 1 to 9426
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   user_id                      9426 non-null   int64         
 1   is_trial                     9426 non-null   object        
 2   first_event_date             9426 non-null   datetime64[ns]
 3   subscription_renewal_amount  9426 non-null   int64         
 4   Subs type                    2 non-null      object        
 5   Trial period_days            1 non-null      float64       
 6   Subscription period_month    2 non-null      float64       
 7   First payment_doll           2 non-null      float64       
 8   Rebill payment__doll         2 non-null      float64       
dtypes: datetime64[ns](1), float64(4), int64(2), object(2)
memory usage: 662.9+ KB


<>:4: SyntaxWarning:

invalid escape sequence '\d'

<>:5: SyntaxWarning:

invalid escape sequence '\d'

<>:4: SyntaxWarning:

invalid escape sequence '\d'

<>:5: SyntaxWarning:

invalid escape sequence '\d'

/tmp/ipython-input-3787985125.py:4: SyntaxWarning:

invalid escape sequence '\d'

/tmp/ipython-input-3787985125.py:5: SyntaxWarning:

invalid escape sequence '\d'

/tmp/ipython-input-3787985125.py:3: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.



In [214]:
df

0     user_id is_trial first_event_date  subscription_renewal_amount  \
1       21889        1       2022-01-01                            0   
2       21890        1       2019-07-01                            0   
3       21891        1       2019-07-01                            0   
4       21892        1       2019-07-01                            0   
5       21894        1       2019-07-01                            0   
...       ...      ...              ...                          ...   
9422    23472        0       2019-07-06                            3   
9423    22991        0       2019-07-05                            7   
9424    23248        0       2019-07-05                           11   
9425    24355        0       2019-07-07                           23   
9426    24769        1       2019-07-07                            1   

0    Subs type  Trial period_days  Subscription period_month  \
1        trial                7.0                        1.0   
2     wo trial                NaN                        3.0   
3          NaN                NaN                        NaN   
4          NaN                NaN                        NaN   
5          NaN                NaN                        NaN   
...        ...                ...                        ...   
9422       NaN                NaN                        NaN   
9423       NaN                NaN                        NaN   
9424       NaN                NaN                        NaN   
9425       NaN                NaN                        NaN   
9426       NaN                NaN                        NaN   

0     First payment_doll  Rebill payment__doll  
1                   6.99                 29.99  
2                  40.00                 40.00  
3                    NaN                   NaN  
4                    NaN                   NaN  
5                    NaN                   NaN  
...                  ...                   ...  
9422                 NaN                   NaN  
9423                 NaN                   NaN  
9424                 NaN                   NaN  
9425                 NaN                   NaN  
9426                 NaN                   NaN  

[9426 rows x 9 columns]

In [215]:
# На початку не думав розюивати на 2 таблиці, але всеж-таки розібю
# Перша таблиця — усі рядки, але тільки перші 4 колонки
df_users = df.iloc[:, :4].copy()
df_users

0     user_id is_trial first_event_date  subscription_renewal_amount
1       21889        1       2022-01-01                            0
2       21890        1       2019-07-01                            0
3       21891        1       2019-07-01                            0
4       21892        1       2019-07-01                            0
5       21894        1       2019-07-01                            0
...       ...      ...              ...                          ...
9422    23472        0       2019-07-06                            3
9423    22991        0       2019-07-05                            7
9424    23248        0       2019-07-05                           11
9425    24355        0       2019-07-07                           23
9426    24769        1       2019-07-07                            1

[9426 rows x 4 columns]

In [216]:
# Друга таблиця — тільки перші 2 рядки, починаючи з 4-ї колонки
df_types = df.iloc[:2, 4:].copy()
df_types

0 Subs type  Trial period_days  Subscription period_month  First payment_doll  \
1     trial                7.0                        1.0                6.99   
2  wo trial                NaN                        3.0               40.00   

0  Rebill payment__doll  
1                 29.99  
2                 40.00

## Завдання_1

Тут, думаю, треба поснити як я думаю. Ті хто взяв пробний період заплатили 6,99 за 7 днів користування сервісом, наступна оплата помісячно з вартістю 29,99

$$
\text{Total revenue}_{trial} = \sum_{i=1}^{N}
\Big( \text{First payment} + \text{Rebill payment} \times \text{subscription_renewal_amount}_i \Big)
$$


In [217]:
first_payment = df_types.iloc[0, 3]
rebill_payment = df_types.iloc[0, 4]

In [218]:
# Порахую кількість людей хто взяв пробну підписку
df_trial = df_users[df_users['is_trial'] == '0'].copy()


In [219]:
# Чисто випадково вирішив перевірити на дублікати
df_trial['user_id'].value_counts().head(10)

user_id
23712    31
21916    26
22008    26
23402    26
23213    26
22022    26
22026    26
22027    26
23724    26
23309    26
Name: count, dtype: int64

In [220]:
# Це вже цікаво, цнікальних корстуваці то тільки 980, тобто з кожним платежем рядок до таблиці додавався
df_trial['user_id'].nunique()

980

In [221]:
df_trial[df_trial['user_id'] == 23712]


0     user_id is_trial first_event_date  subscription_renewal_amount
3660    23712        0       2019-07-06                            1
4338    23712        0       2019-07-06                            2
4833    23712        0       2019-07-06                            3
5241    23712        0       2019-07-06                            4
5242    23712        0       2019-07-06                            5
5612    23712        0       2019-07-06                            6
5946    23712        0       2019-07-06                            7
6233    23712        0       2019-07-06                            8
6500    23712        0       2019-07-06                            9
6752    23712        0       2019-07-06                           10
6753    23712        0       2019-07-06                           10
6754    23712        0       2019-07-06                           10
6755    23712        0       2019-07-06                           10
6756    23712        0       2019-07-06                           10
6757    23712        0       2019-07-06                           10
6990    23712        0       2019-07-06                           11
7212    23712        0       2019-07-06                           12
7404    23712        0       2019-07-06                           13
7592    23712        0       2019-07-06                           14
7760    23712        0       2019-07-06                           15
7918    23712        0       2019-07-06                           16
8068    23712        0       2019-07-06                           17
8202    23712        0       2019-07-06                           18
8336    23712        0       2019-07-06                           19
8469    23712        0       2019-07-06                           20
8589    23712        0       2019-07-06                           21
8704    23712        0       2019-07-06                           22
8807    23712        0       2019-07-06                           23
8906    23712        0       2019-07-06                           24
9003    23712        0       2019-07-06                           25
9105    23712        0       2019-07-06                           26

In [222]:
# Групуємо дані за користувачем і беремо максимум поновлень
df_trial_unique = df_trial.groupby('user_id', as_index=False)['subscription_renewal_amount'].max()
df_trial_unique

user_id  subscription_renewal_amount
0      21889                           11
1      21890                            1
2      21892                            1
3      21896                            2
4      21898                            3
..       ...                          ...
975    25224                            1
976    25227                           26
977    27515                            3
978    47126                            6
979   115217                            1

[980 rows x 2 columns]

In [223]:
# Рахую індивідуальний дохід
df_trial_unique['user_revenue'] = first_payment + rebill_payment * df_trial_unique['subscription_renewal_amount']


In [224]:
df_trial_unique

user_id  subscription_renewal_amount  user_revenue
0      21889                           11        336.88
1      21890                            1         36.98
2      21892                            1         36.98
3      21896                            2         66.97
4      21898                            3         96.96
..       ...                          ...           ...
975    25224                            1         36.98
976    25227                           26        786.73
977    27515                            3         96.96
978    47126                            6        186.93
979   115217                            1         36.98

[980 rows x 3 columns]

In [225]:
# Загальний результат
total_trial_revenue = df_trial_unique['user_revenue'].sum()
n_trial_users = len(df_trial_unique)

print(f"Кількість унікальних trial-користувачів: {n_trial_users}")
print(f"Сукупний дохід від trial-користувачів: ${total_trial_revenue:,.2f}")

Кількість унікальних trial-користувачів: 980
Сукупний дохід від trial-користувачів: $193,627.92


*Висновок*:

Було проведено очищення даних від дублікатів та коректно розраховано індивідуальний і загальний дохід від користувачів, які оформили підписку з пробним періодом.

**Кількість унікальних trial-користувачів: 980**

**Сукупний дохід від trial-користувачів: $193,627.92**

## Завдання 2

### Частина "а"

In [226]:
df_trial['subscription_renewal_amount'].nunique()

26

In [227]:
# Рахуємо кількість користувачів для кожного числа поновлень
renewal_counts = (df_trial_unique['subscription_renewal_amount'].value_counts().sort_index().reset_index())

renewal_counts.columns = ['Number of renewals', 'Number of users']
renewal_counts

Number of renewals  Number of users
0                    1              349
1                    2              134
2                    3               86
3                    4               58
4                    5               32
5                    6               40
6                    7               29
7                    8               16
8                    9               13
9                   10               17
10                  11               16
11                  12               16
12                  13               11
13                  14               16
14                  15                8
15                  16                5
16                  17                8
17                  18                9
18                  19                9
19                  20                6
20                  21                6
21                  22                4
22                  23                5
23                  24                3
24                  25                6
25                  26               78

In [228]:
fig = px.bar(
    renewal_counts,
    x='Number of renewals',
    y='Number of users',
    title='Кількість користувачів за кількістю ребілів (Trial-підписки)',
    color='Number of renewals',
    text='Number of users',
    color_continuous_scale='Viridis'
)

fig.update_traces(texttemplate='%{text}', textposition='outside')
fig.update_layout(
    xaxis_title='Кількість ребілів',
    yaxis_title='Кількість користувачів',
    showlegend=False,
    height=500
)
fig.show()

### Частина "б"

In [229]:
revenue_by_segment = (
    df_trial_unique
    .groupby('subscription_renewal_amount')['user_revenue']
    .sum()
    .reset_index()
)

fig = px.bar(
    revenue_by_segment,
    x='subscription_renewal_amount',
    y='user_revenue',
    color='user_revenue',
    text='user_revenue',
    color_continuous_scale='Plasma',
    title='Сукупний дохід за кількістю ребілів (Trial-підписки)'
)

fig.update_traces(texttemplate='$%{text:.2f}', textposition='outside')
fig.update_layout(
    xaxis_title='Кількість ребілів',
    yaxis_title='Сумарний дохід ($)',
    showlegend=False,
    height=500
)
fig.show()


### Частина "в"

In [230]:
# Перевірив які роки включені в даний датасет
df_trial['year'] = df_trial['first_event_date'].dt.year
df_trial['year'].unique()

array([2019], dtype=int32)

In [231]:
# Перевірка пероду данних
print("Найраніша дата:", df_trial['first_event_date'].min())
print("Найпізніша дата:", df_trial['first_event_date'].max())

Найраніша дата: 2019-07-01 00:00:00
Найпізніша дата: 2019-07-07 00:00:00


In [232]:
# беремо лише останній запис для кожного користувача
df_trial_last = df_trial.sort_values('subscription_renewal_amount').groupby('user_id').tail(1)

# рахуємо правильний дохід
df_trial_last['user_revenue'] = first_payment + rebill_payment * df_trial_last['subscription_renewal_amount']

# групуємо за датою
revenue_timeline = (
    df_trial_last
    .groupby('first_event_date')['user_revenue']
    .sum()
    .reset_index()
    .sort_values('first_event_date')
)

In [233]:
fig = px.line(
    revenue_timeline,
    x='first_event_date',
    y='user_revenue',
    text='user_revenue',
    color_discrete_sequence=['#1f77b4'],
    title='Динаміка доходу у часі (Trial-підписки)',
    markers=True
)

fig.update_traces(
    textposition='top center',   # 👈 підпис над точкою
    texttemplate='%{text:,.0f}$' # 👈 формат — без десяткових, з $
)

fig.update_layout(
    xaxis_title='Дата',
    yaxis_title='Сумарний дохід ($)',
    height=500,
)
fig.show()